gsuite_reports_rules/gsuite_drive_visibility_change.py

```diff
@@ -1,42 +1,123 @@
 from panther_base_helpers import deep_get
 from panther_base_helpers import gsuite_parameter_lookup as param_lookup
 
+EXCLUDED_DOMAINS = ["runpanther.io"]
+
+VISIBILITY = ["people_with_link", "public_on_the_web", "shared_externally", "unknown"]
+
+DOC_TITLE = "<UNKNOWN_TITLE>"
+TARGET_USER_EMAIL = "<UNKNOWN_USER>"
+TARGET_DOMAIN = "<UNKNOWN_DOMAIN>"
+ACCESS_SCOPE = "<UNKNOWN_ACCESS_SCOPE>"
+NEW_VISIBILITY = "<UNKNOWN_VISIBILITY>"
+
 
 def rule(event):
     if deep_get(event, "id", "applicationName") != "drive":
         return False
 
+    #########
+    # for target_domain
+    change_document_visibility = False
+
+    for details in event.get("events", [{}]):
+        if (
+            details.get("type") == "acl_change"
+            and details.get("name") == "change_document_visibility"
+            and param_lookup(details.get("parameters", {}), "new_value") != ["private"]
+            and param_lookup(details.get("parameters", {}), "target_domain") not in EXCLUDED_DOMAINS
+            and param_lookup(details.get("parameters", {}), "visibility") in VISIBILITY
+        ):
+            global TARGET_DOMAIN
+            TARGET_DOMAIN = param_lookup(details.get("parameters", {}), "target_domain")
+            global NEW_VISIBILITY
+            NEW_VISIBILITY = param_lookup(details.get("parameters", {}), "visibility")
+            global DOC_TITLE
+            DOC_TITLE = param_lookup(details.get("parameters", {}), "doc_title")
+
+            change_document_visibility = True
+            break
+
+    # "change_document_access_scope" events are always paired with "change_document_visibility" events
+    # the "target_domain" and "visibility" attributes are the same
+    # the set of a "change_document_access_scope" and a "change_document_visibility" is paired with a
+    # second set of these events
+    # the first pair of events represents the "old" domain and visibility, and the second pair
+    # represents the "new" domain and visibility
+    if change_document_visibility:
+        for details in event.get("events", [{}]):
+            if (
+                details.get("type") == "acl_change"
+                and details.get("name") == "change_document_access_scope"
+                and param_lookup(details.get("parameters", {}), "new_value") != ["none"]
+            ):
+                global ACCESS_SCOPE
+                ACCESS_SCOPE = param_lookup(details.get("parameters", {}), "new_value")
+        return True
+
+    # TODO: confirm that change_user_access events are mutually exclusive with the above events
+    # split this out
+    #########
+    # for target_user
```

Comment:
nit: should this just be
```suggestion
    # split this out for target_user
```

Reply:
Just realized I have a bunch of TODO statements left in there, along with things like this. Will remove (but keep in my notes).